src/partial-assembly.md

```diff
@@ -0,0 +1,174 @@
+subdomain (L-vector) or element (E-vector) matrices and perform the action of
+**A** using matvecs with **P** or **P** and **G**. While these options are
+natural for low-order discretizations, they are not a good fit for high-order
+methods due to the amount of FLOPs needed for their evaluation, as well as the
+memory transfer needed for a matvec.
+
+MFEM's **partial assembly** functionality computes and stores only **D** (or
+portions of it) and evaluates the actions of **P**, **G** and **B** on-the-fly.
+Critically for performance, MFEM takes advantage of the tensor-product
+structure of the degrees of freedom and quadrature points on quadrilateral and
+hexahedral elements to perform the action of **B** without storing it as a
+matrix. Note that the action of **B** is performed elementwise (it corresponds
+to a block-diagonal matrix), and the blocks depend only on the element order
+and reference geometry. Currently, only fixed order and geometry is supported,
+meaning that all the blocks of **B** are identical.
+
+The partial assembly algorithm requires the optimal amount of memory transfers
+(with respect to the polynomial order) and near-optimal FLOPs for operator
+evaluation. It consists of an operator *setup* phase, that evaluates and stores
+**D** and an operator *apply* (evaluation) phase that computes the action of
+**A** on an input vector. When desired, the setup phase may be done as a
+side-effect of evaluating a different operator, such as a nonlinear residual.
+The relative costs of the setup and apply phases are different depending on the
+physics being expressed and the representation of **D**.
+
+### Parallel Decomposition
+
+After the application of each of the first three transition operators, **P**,
+**G** and **B**, the operator evaluation is decoupled on their ranges, so
+**P**, **G** and **B** allow us to "zoom-in" to subdomain, element and
+quadrature point level, ignoring the coupling at higher levels.
+
+Thus, a natural mapping of **A** on a parallel computer is to split the
+**T-vector** over MPI ranks (a non-overlapping decomposition, as is typically
+used for sparse matrices), and then split the rest of the vector types over
+computational devices (CPUs, GPUs, etc.) as indicated by the shaded regions in
+the diagram above.
+
+One of the advantages of the decomposition perspective in these settings is
+that the operators **P**, **G**, **B** and **D** clearly separate the MPI
+parallelism in the operator (**P**) from the unstructured mesh topology
+(**G**), the choice of the finite element space/basis (**B**) and the geometry
+and point-wise physics **D**. These components also naturally fall in different
+classes of numerical algorithms: parallel (multi-device) linear algebra for
+**P**, sparse (on-device) linear algebra for **G**, dense/structured linear
+algebra (tensor contractions) for **B** and parallel point-wise evaluations for
+**D**.
+
```

Comment:
I feel like a brief comment on imposing essential boundary conditions for PA operators would be useful, either here or at the end of the "Finite Element Operator Decomposition" section. The way this is done differs a bit with PA compared to FA, and that has tripped me up a couple times.

Reply:
Good suggestion. I added a short section.